### PR TITLE
Fixes MRESOLVER-7, third try

### PR DESCRIPTION
**Pull request by hwellmann**
Rebased PR #10 on master and added a small change which reverts dependency management to its former somewhat broken state, see MRESOLVER-9.

This fixes failing tests in the `maven` and `maven-integration-testing` projects.

### CODE REVIEW

**Comment by slachiewicz** [comment]
During verification, I found that one change is being rolled out - https://github.com/apache/maven-resolver/commit/a245b56afd479640ce12a3cd047b463a629bca0f I do not think it was intentional. 
You can also freely squash my commit with WorkerThreadFactory

**Comment by hwellmann** [comment]
I've reintroduced the change made in a245b56 that got lost during rebase - thanks for spotting this!

What exactly did you mean regarding the "commit with WorkerThreadFactory"?

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/DefaultDependencyCollector.java [hunk 1]
```diff
@@ -666,46 +679,4 @@ private static boolean isLackingDescriptor( Artifact artifact )
         }
         return repositories;
     }
-
-    private static List<? extends Version> filterVersions( Dependency dependency, VersionRangeResult rangeResult,
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/DefaultDependencyCollector.java]: Why has this method been removed in this commit?

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/DefaultDependencyCollector.java [hunk 2]
```diff
@@ -89,6 +90,10 @@
 
     static final String CONFIG_PROP_MAX_CYCLES = "aether.dependencyCollector.maxCycles";
 
+    private static final String CONFIG_PROP_NUM_ARTIFACT_DESCRIPTOR_THREADS = "aether.artifact.descriptor.threads";
+    private static final String CONFIG_PROP_NUM_ARTIFACT_DESCRIPTOR_THREADS_ALT = "maven.artifact.descriptor.threads";
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/DefaultDependencyCollector.java]: I don't see a need for an alternative. As soon as the package names will change we will likely rename properties too.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/PremanagedDependency.java [hunk 3]
```diff
@@ -35,18 +42,39 @@
 
     final Boolean premanagedOptional;
 
+    /**
+     * @since 1.1.0
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/PremanagedDependency.java]: Please change to 1.4.0.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/PremanagedDependency.java [hunk 4]
```diff
@@ -35,18 +42,39 @@
...
+     */
+    final Collection<Exclusion> premanagedExclusions;
+
+    /**
+     * @since 1.1.0
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/PremanagedDependency.java]: Please change to 1.4.0.

**Comment by michael-o** [comment]
All ITs pass for me. As soon as the requested changes are performed I will merge into master.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/DefaultDependencyCollector.java (hunk 1 above)
Comment (hwellmann) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/DefaultDependencyCollector.java]: It has been moved to `DependencyCollectionUtils`.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/DefaultDependencyCollector.java (hunk 2 above)
Comment (hwellmann) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/DefaultDependencyCollector.java]: Ok, removed.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/PremanagedDependency.java (hunk 3 above)
Comment (hwellmann) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/PremanagedDependency.java]: Done.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/PremanagedDependency.java (hunk 4 above)
Comment (hwellmann) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/PremanagedDependency.java]: Done.

**Comment by michael-o** [comment]
I have now pushed a branch with minimal changes. Everyone's invited to have look. If I hear no objections, I'll merge in a couple of days.

**Comment by nilson-aguiar** [comment]
@michael-o this is not going to be merged?

**Comment by michael-o** [comment]
??? This has been merged two months ago.

**Comment by nilson-aguiar** [comment]
Sorry then, saw it as closed instead of merged then i asked. Thanks for the reply.

**Comment by sschuberth** [comment]
Does this new feature imply some API usage changes? Because in our [OSS Review Toolkit](https://github.com/heremaps/oss-review-toolkit) (which analyzes project dependencies) we actually see some POMs only being downloaded after we expect them to be present. Downgrading to version 1.3.3 resolves the issue. For details please see https://github.com/heremaps/oss-review-toolkit/pull/2113.

**Comment by michael-o** [comment]
@sschuberth This has been reverted because it suffered from race conditions.

**Comment by sschuberth** [comment]
Thanks @michael-o, I was just reading about that in [MRESOLVER-92](https://issues.apache.org/jira/browse/MRESOLVER-92) which is mentioned in the [1.4.1 release announcement](https://www.mail-archive.com/users@maven.apache.org/msg141606.html). However, interestingly the issue we're seeing occurs with both versions 1.4.0 and 1.4.1. So maybe it's something that's not caused by the parallel download of POMs after all.

**Comment by michael-o** [comment]
@sschuberth  If you see some other error, please report.

**Comment by sschuberth** [comment]
@michael-o, in our case, the root cause turned out to be a side effect of https://www.alphabot.com/security/blog/2020/java/Your-Java-builds-might-break-starting-January-13th.html.

**Comment by Eskibear** [comment]
Any update? Will you guys re-enable this improvement to download pom files in parallel?

**Comment by michael-o** [comment]
> 
> 
> Any update? Will you guys re-enable this improvement to download pom files in parallel?

No, unless someone is willing to work on the race conditions.

**Comment by Eskibear** [comment]
All I can find about the race conditions is @Tibor17 's [comments in MRESOLVER-7](https://issues.apache.org/jira/browse/MRESOLVER-7?focusedCommentId=16885618&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16885618) . And it seems that details were discussed in the ASF slack but it is internal.  So it might be hard for **someone** to work on it without knowing the details.

@hwellmann @slachiewicz May I ask that, are you still working on the fix, or some other ASF guys will take it over?

**Comment by michael-o** [comment]
@Eskibear, no, no one is working on it. We prefer stability over performance.

**Comment by Eskibear** [comment]
@michael-o thanks for the information. BTW, could you share some details (or point out the link) about the failed cases, race conditions mentioned above. My point is, if someone from community is willing to fix that, your previous discussion would help a lot.

**Comment by michael-o** [comment]
Basically, we have frequent race conditions in the Maven ITs. @Tibor17 Do you remember the discussion the mailing list? I guess it was on Slack only :-(

@Eskibear. I think you can reproduce it yourself. Fetch the first broken Resolver release and run Maven ITs multiple times. I think you should see failures not related to Maven itself.

**Comment by Tibor17** [comment]
@michael-o 
@Eskibear 
Hello guys. You'r right! We should restart the work.
Can we start again during the weekend?

**Comment by Eskibear** [comment]
Any news or follow-ups?

**Comment by michael-o** [comment]
@Eskibear, since we have other serious issues with concurency this isn't happening soon.

**Comment by Tibor17** [comment]
@michael-o @Eskibear 
If you have time these days or weeks we can start analysing the code @hwellmann was working on.
We should define some order of tasks.
Lets start with the first analysis where we will find the parts where the downloads can be made parallel.

**Comment by michael-o** [comment]
@michael-o No, I won't. Looking at https://issues.apache.org/jira/browse/MRESOLVER-114 will make things worse because we do not have decent concurrency support. Even if we solve this one, the synchronization will trick us anyway.

**Comment by Eskibear** [comment]
@michael-o I just glanced at MRESOLVER-114, and I don't think it's related with MRESOLVER-7. 
Learning for the [comments](https://issues.apache.org/jira/browse/MRESOLVER-114?focusedCommentId=17127769&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17127769), MRESOLVER-114 is about resolving artifacts in parallel. 
While here in MRESOLVER-7, the problem is, when the list of dependencies is already known, it performs **downloading** of the corresponding pom files in a single thread. Given you guys have more context information, correct me if I'm wrong. 
What I see is, networking is the bottleneck. It would be a great performance boost even if we just make the downloading parallel. 


> Lets start with the first analysis where we will find the parts where the downloads can be made parallel.

@Tibor17 Agree with you that we should do some first analysis. Below is what I find in master branch, where downloading dependencies is mono-threaded. As far as I learn, basic idea of this PR was to make below logic parallel. 
https://github.com/apache/maven-resolver/blob/86c1cb28042d2e751ddd657a7490c997cff4cd26/maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/DefaultDependencyCollector.java#L347-L351

**Comment by michael-o** [comment]
Please move discussion to the ticket because this PR has been closed.

**Comment by Tibor17** [comment]
@Eskibear 
We have to find the place where the files are physically downloaded, because it is the place which should run in parallel. The caller methods should not go to the thread pool because Maven's implementations of objects are not thread safe. I do not want to make al of them immutable. And why to call the Java mess in forked thread. It is not a performance penalty. The penatly we pay is the physical download.

**Comment by Tibor17** [comment]
@Eskibear 
If I am right, [here](https://github.com/apache/maven-resolver/blob/master/maven-resolver-transport-http/src/main/java/org/eclipse/aether/transport/http/HttpTransporter.java#L290) is one of the low level places. I do not like shared objects on instance and here is the shared [state](https://github.com/apache/maven-resolver/blob/master/maven-resolver-transport-http/src/main/java/org/eclipse/aether/transport/http/HttpTransporter.java#L105).

**Comment by michael-o** [comment]
Maven does *not* use the HTTP Transporter. We use Resolver via Wagon only.

**Comment by jira-importer** [comment]
Resolve #1080

**Comment by jira-importer** [comment]
Resolve #817

**Pull request closed without merging**